feat(examples): Add kubectl-friendly YAML examples for TrainJob and TrainingRuntime

NarayanaSabari opened the pull request:
## Description

This PR adds comprehensive kubectl-friendly YAML examples for Kubeflow Trainer, addressing the need for simple, out-of-the-box examples that can be applied directly with `kubectl`.

Fixes #2770

## Motivation

As noted in #2770, while Jupyter notebook examples are great for AI practitioners, platform administrators and users who prefer working with kubectl need ready-to-use YAML examples. The existing operator guides have embedded YAML snippets, but users were looking for standalone examples in the repository that:
- Work out of the box
- Don't require GPU resources
- Show both basic and advanced configurations
- Can be applied directly with `kubectl apply -f`

This is especially important for "GPU-poor" users who want to explore and learn Kubeflow Trainer.

## Changes

Added a new `examples/yaml/` directory with **8 ready-to-use examples** organized by complexity:

### Basic Examples (`examples/yaml/basic/`)
1. **`01-hello-world.yaml`** - Simplest TrainJob with echo command (no GPU)
2. **`02-multi-node.yaml`** - Multi-node distributed training simulation (no GPU)
3. **`03-with-runtime.yaml`** - Custom namespace-scoped TrainingRuntime creation and usage (no GPU)
4. **`04-pytorch-simple.yaml`** - Real PyTorch MNIST training (GPU optional, uses public image)

### Advanced Examples (`examples/yaml/advanced/`)
1. **`01-podspec-overrides.yaml`** - Comprehensive pod customization with `podTemplateOverrides`
2. **`02-kueue-integration.yaml`** - Queue-based job scheduling with Kueue
3. **`03-volcano-integration.yaml`** - Gang scheduling for distributed training with Volcano
4. **`04-multi-step.yaml`** - Multi-step pipeline with dataset initialization

### Documentation
- **`examples/yaml/README.md`** - Main guide with quick start, common commands, troubleshooting
- **`examples/yaml/basic/README.md`** - Detailed guide for basic examples
- **`examples/yaml/advanced/README.md`** - Production best practices and advanced patterns
- Updated **`examples/README.md`** - Added clear separation between YAML (Platform Admins) and SDK (AI Practitioners) examples

## Key Features

✅ **All examples verified against actual CRD schemas** (not just documentation)
- Uses correct `podTemplateOverrides` API structure with `targetJobs` array
- Uses correct `initializer.dataset` API with `storageUri` field
- No deprecated or non-existent fields

✅ **GPU-poor friendly**
- 3 out of 4 basic examples require NO GPU
- PyTorch example works with or without GPU (auto-detects)

✅ **Production-ready**
- Uses publicly available images (e.g., `docker.io/kubeflowkatib/pytorch-mnist:v1beta1-45c5727`)
- Includes best practices for resource limits, security, scheduling
- Shows integration with Kueue and Volcano

✅ **Complete kubectl workflows**
- Every example includes apply/check/logs/delete commands
- Comprehensive troubleshooting section
- Common debugging patterns

## Testing

All YAML files have been validated against:
- TrainJob CRD schema (`trainer.kubeflow.org/v1alpha1`)
- TrainingRuntime CRD schema (namespace-scoped)
- ClusterTrainingRuntime CRD schema (cluster-scoped)

Images verified as publicly available:
- `docker.io/kubeflowkatib/pytorch-mnist:v1beta1-45c5727` ✅
- `python:3.11-slim` ✅

## Notes

- These examples complement (not replace) the Python SDK examples
- They're specifically designed for platform administrators and kubectl users
- All examples can be applied immediately without building custom images
- Documentation clearly separates use cases: YAML for Platform Admins, SDK for AI Practitioners

## Related Issues

- Addresses feedback from @kannon92 about wanting simple hello-world examples without requiring GPU
- Implements suggestions from @andreyvelich about showing PodSpecOverrides usage
- Follows @kramaranya's idea of complete kubectl workflows for Platform Admins

cc @kannon92 @andreyvelich @kramaranya @astefanutti @xigang

NarayanaSabari committed:
added yaml examples for trainer

Signed-off-by: narayanasabari <sabarinarayanakg@proton.me>

kannon92 commented:
This is exactly what I wanted!

I haven't verified if this actually work though. It would be worth doing that just to make sure.

kannon92 commented on examples/yaml/advanced/02-kueue-integration.yaml:
```diff
@@ -0,0 +1,73 @@
+# Advanced Example 2: Kueue Integration
```
I'm not sure this is correct actually.

AFAIK TrainJob Integration for Kueue should only require adding a label to the top level TrainJob object.

PodIntegrationOverrides were added for more advancued features for kueue like TopoloyAwareScheduling.

NarayanaSabari committed:
Examples fixed

Signed-off-by: narayanasabari <sabarinarayanakg@proton.me>

NarayanaSabari commented:
@kannon92 You're absolutely right! Thank you for catching that. 

I've updated the Kueue example to use the correct approach:
- Moved the `kueue.x-k8s.io/queue-name` label to the TrainJob metadata level
- Removed the unnecessary `podTemplateOverrides` for basic Kueue integration

I also took the opportunity to:
- Fix the Volcano example to use the official `podGroupPolicy` API
- Fix the multi-step example where `initializer` was incorrectly nested under `trainer`

All examples now match the actual CRD API definitions. Thanks for the detailed feedback!

andreyvelich commented on examples/yaml/basic/02-multi-node.yaml:
```diff
@@ -0,0 +1,59 @@
+        echo "Rank: ${RANK:-0}"
+        echo "World Size: ${WORLD_SIZE:-1}"
+        echo "Master Addr: ${MASTER_ADDR:-localhost}"
+        echo "Master Port: ${MASTER_PORT:-29500}"
```
That might require PET_ prefix, did you run this example locally?

andreyvelich commented on examples/yaml/advanced/02-kueue-integration.yaml:
```diff
@@ -0,0 +1,65 @@
+# Advanced Example 2: Kueue Integration
```
@kannon92 @mimowo @tenzen-y @kaisoz Any thoughts on this Kueue and TrainJob example?

andreyvelich reviewed (commented):
Thanks @NarayanaSabari, I left a few thoughts.

NarayanaSabari commented on examples/yaml/advanced/03-volcano-integration.yaml:
```diff
@@ -0,0 +1,61 @@
+  namespace: default
+  labels:
+    app: training
+    scheduler: volcano
```
Removed the unnecessary labels. The rewritten example now only has what's needed: the `ClusterTrainingRuntime` with `podGroupPolicy` and a minimal `TrainJob` referencing it.

NarayanaSabari committed:
fix(examples): address maintainer feedback on YAML examples

- Remove hello-world, with-runtime, pytorch-simple examples
- Fix multi-node: remove custom image, use PET_* env var prefix
- Fix volcano: move podGroupPolicy to ClusterTrainingRuntime
- Fix podspec-overrides: move podTemplateOverrides to top-level,
  use trainer.resourcesPerNode and trainer.env instead of invalid
  ContainerOverride fields
- Trim all READMEs to avoid duplicating kubeflow.org docs

Signed-off-by: Sabari Narayana <sabarinarayanakg@proton.me>

andreyvelich reviewed (commented):
@NarayanaSabari Please could you rebase your PR, and update the examples to use RuntimePatches API?

NarayanaSabari committed:
feat(examples): migrate YAML examples to RuntimePatches API

Address review feedback from @andreyvelich on PR #2925:

- Rename 01-podspec-overrides.yaml -> 01-runtime-patches.yaml and rewrite
  using the new spec.runtimePatches API (PodTemplateOverrides was removed
  in favor of the manager-keyed RuntimePatches mechanism).
- Inline a TrainingRuntime in 04-multi-step.yaml so the dataset-initializer
  step is actually defined; the default torch-distributed runtime has no
  initializer replicatedJob and the example would fail validation otherwise.
- Replace nonexistent trainer.kubeflow.org/job-name selectors throughout
  with jobset.sigs.k8s.io/jobset-name (for pods/podgroups/workloads) and
  trainer.kubeflow.org/trainjob-ancestor-step (to target a specific step).
- Bump default PyTorch image to 2.10.0 to match upstream torch-distributed
  ClusterTrainingRuntime.
- Add ManagedBy hint for MultiKueue in the Kueue example.
- Trim READMEs to avoid duplicating kubeflow.org documentation.

Signed-off-by: Sabari Narayana <sabarinarayanakg@proton.me>

andreyvelich commented on examples/yaml/basic/01-multi-node.yaml:
```diff
@@ -0,0 +1,50 @@
+# Example 1: Multi-Node Distributed Training
```
I am not sure if we need to have subfolders for basic and advanced example, since we have only one `multi-node` example. Let's keep all of them under `examples/yaml` subfolder.

andreyvelich commented on examples/yaml/basic/01-multi-node.yaml:
```diff
@@ -0,0 +1,50 @@
+spec:
+  runtimeRef:
+    apiGroup: trainer.kubeflow.org
+    kind: ClusterTrainingRuntime
```
This is not needed.

```suggestion
```

andreyvelich commented on examples/yaml/basic/01-multi-node.yaml:
```diff
@@ -0,0 +1,50 @@
+  trainer:
+    numNodes: 3
+    command:
+      - python3
```
Why not use `torchrun` as an entrypoint?

andreyvelich commented on examples/yaml/basic/01-multi-node.yaml:
```diff
@@ -0,0 +1,50 @@
+        for i in range(1, 6):
+            print(f"Epoch {i}: batch on node rank {node_rank}")
+            time.sleep(2)
+        print(f"Training complete on node rank {node_rank}")
```
You can check example from here to use `torch()` API to get info about env: https://www.kubeflow.org/docs/components/trainer/user-guides/pytorch/#pytorch-distributed-environment
E.g.
```py
import os
import torch
import torch.distributed as dist

device, backend = ("cuda", "nccl") if torch.cuda.is_available() else ("cpu", "gloo")
dist.init_process_group(backend=backend)

print("PyTorch Distributed Environment")
print(f"Using device: {device}")
print(f"WORLD_SIZE: {dist.get_world_size()}")
print(f"RANK: {dist.get_rank()}")
print(f"LOCAL_RANK: {os.environ['LOCAL_RANK']}")
```

andreyvelich commented on examples/yaml/02-runtime-patches.yaml:
```diff
@@ -0,0 +1,84 @@
+  namespace: default
+spec:
+  runtimeRef:
+    apiGroup: trainer.kubeflow.org
+    kind: ClusterTrainingRuntime
```
same here.

andreyvelich commented on examples/yaml/advanced/01-runtime-patches.yaml:
```diff
@@ -0,0 +1,84 @@
+kind: TrainJob
+metadata:
+  name: runtime-patches-example
+  namespace: default
```
You don't need to set `default` namespace. Feel free to update other YAMLs too.

```suggestion
```

andreyvelich commented on examples/README.md:
```diff
@@ -1,5 +1,59 @@
+├── pytorch/       # PyTorch SDK examples (AI Practitioners)
+├── deepspeed/     # DeepSpeed framework examples
+├── mlx/           # MLX framework examples
+└── torchtune/     # TorchTune fine-tuning examples
```
Can we move `yaml` examples at the bottom, so users read SDK example first?

```suggestion
examples/
├── pytorch/       # PyTorch SDK examples (AI Practitioners)
├── deepspeed/     # DeepSpeed framework examples
├── mlx/           # MLX framework examples
└── torchtune/     # TorchTune fine-tuning examples
├── yaml/          # YAML examples for kubectl users (Platform Admins)
```

andreyvelich commented on examples/README.md:
````diff
@@ -1,5 +1,59 @@
+## Directory Structure
+
+```
+examples/
````
Please also add more example folders (e.g. local, megatron, flux, xgboost)

andreyvelich commented on examples/README.md:
````diff
@@ -1,5 +1,59 @@
+└── torchtune/     # TorchTune fine-tuning examples
+```
+
+## For Platform Administrators (YAML + kubectl)
````
Move this after  For AI Practitioners section

andreyvelich commented on examples/README.md:
```diff
@@ -1,5 +1,59 @@
+    num_nodes=2,
+    entrypoint=train,
+)
+train_job.create()
```
This need to be updated. See this: https://github.com/kubeflow/sdk#run-your-first-pytorch-distributed-job

andreyvelich commented on examples/README.md:
```diff
@@ -1,5 +1,59 @@
+- [Kubeflow Trainer Documentation](https://www.kubeflow.org/docs/components/trainer/)
+- [Getting Started Guide](https://www.kubeflow.org/docs/components/trainer/getting-started/)
+- [Runtime Guide](https://www.kubeflow.org/docs/components/trainer/operator-guides/runtime/)
+
```
Let's also link sdk docs: https://sdk.kubeflow.org/en/latest/

andreyvelich reviewed (commented):
Thanks @NarayanaSabari, overall lgtm, I left a few comments.
/assign @robert-bell @akshaychitneni @astefanutti @tenzen-y @Fiona-Waters @kramaranya @abhijeet-dhumal

NarayanaSabari committed:
docs(examples): address review feedback on YAML examples

Address review feedback from @andreyvelich on PR #2925:

- Flatten examples/yaml/: drop the basic/ and advanced/ subfolders since
  there is only one "basic" example. All YAMLs now live directly under
  examples/yaml/ and are renumbered 01-05.
- Remove namespace: default from all manifests (it is the implicit default
  and adds noise).
- Use torchrun as the entrypoint in 01-multi-node.yaml and 02-runtime-patches.yaml,
  with torch.distributed API in the script (matches the pattern in the
  PyTorch user guide).
- Restructure examples/README.md: SDK section first, YAML section last;
  list all framework subfolders (megatron, flux, xgboost, jax, local);
  refresh SDK snippet from kubeflow/sdk README; link sdk.kubeflow.org.

Signed-off-by: Sabari Narayana <sabarinarayanakg@proton.me>

andreyvelich reviewed (commented):
Looks great @NarayanaSabari!
Thanks for your contributions!
/lgtm
/approve
